### PR TITLE
Gripper real

### PR DESCRIPTION
**Pull request by GaiaZano05**
שהחיינו וקיימנו והגיענו

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java [hunk 1]
```diff
@@ -1,18 +1,77 @@
...
+import edu.wpi.first.units.Units;
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.DutyCycleEncoder;
+import lib.Utils;
 
 public class GripperIOReal implements GripperIO {
+    private final TalonFX angleMotor;
+    private final CANSparkMax rollerMotor;
+    private final DigitalInput sensor;
+    private final DutyCycleEncoder absoluteEncoder = new DutyCycleEncoder(0);
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java:20]: Add an input for the encoder position.

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java [hunk 2]
```diff
@@ -16,17 +23,28 @@ public class GripperConstants {
...
+    public static final int CURRENT_LIMIT = 40;
+    public static final LoggedTunableNumber ABSOLUTE_ENCODER_OFFSET =
+            new LoggedTunableNumber("Gripper/EncoderOffset");
 
     public static final LoggedTunableNumber KP = new LoggedTunableNumber("Gripper/kP");
     public static final LoggedTunableNumber KI = new LoggedTunableNumber("Gripper/kI");
     public static final LoggedTunableNumber KD = new LoggedTunableNumber("Gripper/kD");
+    public static final LoggedTunableNumber KV = new LoggedTunableNumber("Gripper/kD");
+    public static final LoggedTunableNumber KA = new LoggedTunableNumber("Gripper/kD");
+    public static final LoggedTunableNumber KG = new LoggedTunableNumber("Gripper/kD");
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/gripper/GripperConstants.java]: If you're using kG you need to configure `Gravitytype`.

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java [hunk 3]
```diff
@@ -1,18 +1,77 @@
...
+    public void setAngleMotorPower(double power) {
+        angleMotor.setControl(powerRequest.withOutput(power));
+    }
 
     @Override
-    public void setAngle(MutableMeasure<Angle> angle) {}
+    public void setAngle(MutableMeasure<Angle> angle) {
+        angleMotor.setControl(
+                positionRequest.withPosition(
+                        Utils.normalize(new Rotation2d(angle)).getRotations()));
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: Use IEEEEEEEEEEEEEEEE remainder and not normalize, as done in the gripper sim.

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java (hunk 1 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java:20]: 3d71c0d9a9065263a0852481c2a35439f455bb6a

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java (hunk 2 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperConstants.java]: eac9c751b88757ccb73f0fd3260eeba16616adbe

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java (hunk 3 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: f89789cf9df09ea8fc3fbb7537dbec27a1107dee

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java [hunk 4]
```diff
@@ -73,10 +73,11 @@ public Command setWristPosition(MutableMeasure<Angle> angle) {
...
     @Override
     public void periodic() {
-        MutableMeasure<Distance> gripperHeight = elevatorInputs
-                .gripperHeight
-                .mutableCopy()
-                .mut_plus(GripperConstants.GRIPPER_POSITION_z);
+        MutableMeasure<Distance> gripperHeight =
+                elevatorInputs
+                        .gripperHeight
+                        .mutableCopy()
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/gripper/Gripper.java]: Do you have to create a copy here?

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java [hunk 5]
```diff
@@ -16,17 +24,28 @@ public class GripperConstants {
...
     public static final LoggedTunableNumber KD = new LoggedTunableNumber("Gripper/kD");
+    public static final LoggedTunableNumber KV = new LoggedTunableNumber("Gripper/kV");
+    public static final LoggedTunableNumber KA = new LoggedTunableNumber("Gripper/kA");
+    public static final LoggedTunableNumber KG = new LoggedTunableNumber("Gripper/kG");
 
     public static void initConstants() {
         switch (Constants.CURRENT_MODE) {
             case REAL:
-                KP.initDefault(0);
+                KP.initDefault(0.2);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/gripper/GripperConstants.java]: Better to set all gains to zero before you start tuning.

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java [hunk 6]
```diff
@@ -16,17 +24,28 @@ public class GripperConstants {
     public static final Measure<Distance> GRIPPER_POSITION_Y = Units.Meters.of(0);
     public static final Measure<Distance> GRIPPER_POSITION_z = Units.Meters.of(0.6461);
     public static final double ANGLE_MOTOR_GEAR_RATIO = 58.5;
+    public static final InvertedValue ANGLE_INVERTED_VALUE = InvertedValue.Clockwise_Positive;
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:24]: Like in the elevator, no need for these constants

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java [hunk 7]
```diff
@@ -36,5 +55,23 @@ public static void initConstants() {
...
+                .withFeedback(new FeedbackConfigs().withSensorToMechanismRatio(GEAR_RATIO))
+                .withSlot0(
+                        new Slot0Configs()
+                                .withKP(GripperConstants.KP.get())
+                                .withKI(GripperConstants.KI.get())
+                                .withKD(GripperConstants.KD.get())
+                                .withKV(GripperConstants.KV.get())
+                                .withKA(GripperConstants.KA.get())
+                                .withKG(GripperConstants.KG.get())
+                                .withGravityType(GravityTypeValue.Arm_Cosine))
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:72]: Do you set the zero angle to where the docs ask it to be?

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java [hunk 8]
```diff
@@ -1,18 +1,77 @@
...
+import edu.wpi.first.wpilibj.DutyCycleEncoder;
 
 public class GripperIOReal implements GripperIO {
+    private final TalonFX angleMotor;
+    private final CANSparkMax rollerMotor;
+    private final DigitalInput sensor;
+    private final DutyCycleEncoder absoluteEncoder = new DutyCycleEncoder(0);
+    private final MotionMagicExpoTorqueCurrentFOC positionRequest =
+            new MotionMagicExpoTorqueCurrentFOC(0);
+    private final DutyCycleOut powerRequest = new DutyCycleOut(0);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: Unless you wish to run the motor at 100%, using voltage control is more consistent. Also, should use FOC

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java [hunk 9]
```diff
@@ -1,18 +1,77 @@
...
+    public void setAngleMotorPower(double power) {
+        angleMotor.setControl(powerRequest.withOutput(power));
+    }
 
     @Override
-    public void setAngle(MutableMeasure<Angle> angle) {}
+    public void setAngle(MutableMeasure<Angle> angle) {
+        angleMotor.setControl(
+                positionRequest.withPosition(
+                        Math.IEEEremainder(new Rotation2d(angle).getRotations(), Math.PI)));
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: Should π be the divisior?

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java [hunk 10]
```diff
@@ -36,5 +55,23 @@ public static void initConstants() {
                 KD.initDefault(0);
                 break;
         }
+
+        MOTOR_CONFIGURATION
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:56]: Missing Motion Magic Expo configs

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java [hunk 11]
```diff
@@ -1,18 +1,77 @@
...
+    }
+
+    public boolean hasNote() {
+        return sensor.get();
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: Should use some kind of filter here: https://docs.wpilib.org/en/stable/docs/software/advanced-controls/filters/

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java (hunk 7 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:72]: We will

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java (hunk 4 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/Gripper.java]: dbf581f9ce30af978ea3327f1906c3572058f652

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java (hunk 5 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperConstants.java]: 4e5515ea266ed34157bb27a7ad28a7b7dae72a02

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java (hunk 6 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:24]: 6a6d14645ec3b2560832188fd593e35f5a7736b3

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java (hunk 8 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java:23]: f103842388141f1b486cb375dfaa41cee279318e

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java (hunk 9 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: ad2fe549ee24f67785de5e87016ed9e0b1c8df4f

Comment (katzuv) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: Isn't 180 for degrees?

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java (hunk 8 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: The line I commented on is the `powerRequest` decleration. For closed-loop `MotionMagicExpoTorqueCurrentFOC` is indeed the most accurate.

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java (hunk 6 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:24]: I think the highest amperage breaker will trip at 40A

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java (hunk 10 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:56]: e67952c29b3e69f6de236053f62ee2cb0848fb2c

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java (hunk 8 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: 11761ce243ec932a159b68bda6e3b2a085b92d8b

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java (hunk 10 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:56]: Again, no need for additional constants, unless they need to be tunable, which is a reasonable use case here.

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java (hunk 6 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:24]: The limit isn't really a hard limit, and neither is the amperage breaker. From our experience some motors hold 50 amps for about a half second.

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java (hunk 11 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: Not sure how to use it on a digital sensor

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java (hunk 10 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:56]: fb1cace7c0294af728136a2cca0a13a221e35480

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java (hunk 6 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:24]: I see. In this case, you should configure the time threshold after which the motor controller will enforce the supply current limit.

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java (hunk 11 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: Did you go through all of the filter types? One of them works with boolean values.

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java (hunk 9 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: f98a965660c7a011edd08ca4ad470dba5e482cc2

my bad

File: src/main/java/frc/robot/subsystems/gripper/GripperConstants.java (hunk 6 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:24]: f877d97dad5526b7346a04c789b54d2ed57702e1

Comment (katzuv) [reply src/main/java/frc/robot/subsystems/gripper/GripperConstants.java:24]: Why?

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java (hunk 9 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: All good 😊
Commit message could have been better tho

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java [hunk 12]
```diff
@@ -1,18 +1,82 @@
...
+    }
 
     @Override
-    public void updateInputs() {}
+    public void updateInputs() {
+        inputs.angleMotorVoltage.mut_replace(angleMotor.get(), Units.Volts);
+        inputs.rollerMotorVoltage.mut_replace(rollerMotor.get(), Units.Volts);
+        inputs.currentAngle.mut_replace(angleMotor.getPosition().getValue(), Units.Rotations);
+        inputs.hasNote = hasNote();
+        inputs.encoderPosition.mut_replace(absoluteEncoder.getAbsolutePosition(), Units.Degrees);
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: Degrees?

Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: 9ac0471975630b353f34bf00de741cf982cecd3a

File: src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java (hunk 11 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/subsystems/gripper/GripperIOReal.java]: 9ac0471975630b353f34bf00de741cf982cecd3a

**Review by katzuv — APPROVED** [review]

**Pull request merged**